quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -74,15 +85,102 @@ impl super::Endpoint for Endpoint {
                 .with_connection_info(info)
                 .build()
                 .unwrap(),
-            Receiver(()),
+            Receiver::builder()
+                .with_capacity(self.recv_queue_capacity)
+                .build()
+                .unwrap(),
         )
     }
 }
 
-pub struct Receiver(());
+pub struct Receiver {
+    pub queue: VecDeque<Bytes>,
+    capacity: usize,
+    waker: Option<Waker>,
+}
+
+impl Receiver {
+    /// Creates a builder for the default datagram receiver
+    fn builder() -> ReceiverBuilder {
+        ReceiverBuilder::default()
+    }
+
+    /// Returns a datagram if there are any on the queue
+    pub fn recv_datagram(&mut self) -> Option<Bytes> {
+        self.queue.pop_front()
+    }
+
+    /// Dequeues a datagram received from the peer.
+    ///
+    /// # Return value
+    ///
+    /// The function returns:
+    ///
+    /// - `Poll::Pending` if there are no datagrams to be received on the queue. In this case,
+    ///   the caller should retry receiving after the [`Waker`](core::task::Waker) on the provided
+    ///   [`Context`](core::task::Context) is notified.
+    /// - `Poll::Ready(Datagram)` if there exists a datagram to be received.
+    pub fn poll_recv_datagram(&mut self, cx: &mut Context) -> Poll<Bytes> {
+        if self.queue.is_empty() {
+            self.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+        // Since we know the queue is not empty, we can safely unwrap the popped value
+        Poll::Ready(self.queue.pop_front().unwrap())
```

Comment:
Better to match on the pop_front and that way you can avoid `unwrap`

```suggestion
        if let Some(datagram) = self.queue.pop_front() {
            Poll::Ready(datagram)
        } else {
            self.waker = Some(cx.waker().clone());
            Poll::Pending
        }
```